File: src/inline_diff.zig
const std = @import("std");
const tui_text = @import("tui_text.zig");

pub const Range = struct {
    start: usize,
    end: usize,
};

pub const PairRanges = struct {
    old: []Range,
    new: []Range,

    pub fn deinit(self: *PairRanges, allocator: std.mem.Allocator) void {
        allocator.free(self.old);
        allocator.free(self.new);
    }
};

const TokenKind = enum {
    whitespace,
    word,
    punct,
};

const Token = struct {
    start: usize,
    end: usize,
    kind: TokenKind,
};

const max_lcs_tokens: usize = 240;
const max_lcs_cells: usize = 24_000;

pub fn diffRanges(allocator: std.mem.Allocator, old_text: []const u8, new_text: []const u8) !PairRanges {
    if (std.mem.eql(u8, old_text, new_text)) {
        return .{
            .old = try allocator.alloc(Range, 0),
            .new = try allocator.alloc(Range, 0),
        };
    }

    const old_tokens = try tokenize(allocator, old_text);
    defer allocator.free(old_tokens);
    const new_tokens = try tokenize(allocator, new_text);
    defer allocator.free(new_tokens);

    if (old_tokens.len == 0 or new_tokens.len == 0 or
        old_tokens.len > max_lcs_tokens or
        new_tokens.len > max_lcs_tokens or
        old_tokens.len * new_tokens.len > max_lcs_cells)
    {
        return affixRanges(allocator, old_text, new_text);
    }

    const old_matched = try allocator.alloc(bool, old_tokens.len);
    defer allocator.free(old_matched);
    @memset(old_matched, false);
    const new_matched = try allocator.alloc(bool, new_tokens.len);
    defer allocator.free(new_matched);
    @memset(new_matched, false);

    try markLcs(allocator, old_text, old_tokens, new_text, new_tokens, old_matched, new_matched);

    const old_ranges = try rangesFromTokens(allocator, old_tokens, old_matched);
    errdefer allocator.free(old_ranges);
    const new_ranges = try rangesFromTokens(allocator, new_tokens, new_matched);

    return .{
        .old = old_ranges,
        .new = new_ranges,
    };
}

fn tokenize(allocator: std.mem.Allocator, text: []const u8) ![]Token {
    var tokens: std.ArrayList(Token) = .empty;
    errdefer tokens.deinit(allocator);

    var i: usize = 0;
    while (i < text.len) {
        const start = i;
        if (isWhitespace(text[i])) {
            while (i < text.len and isWhitespace(text[i])) : (i += 1) {}
            try tokens.append(allocator, .{ .start = start, .end = i, .kind = .whitespace });
            continue;
        }
        if (isWordByte(text[i])) {
            while (i < text.len and isWordByte(text[i])) : (i += 1) {}
            try tokens.append(allocator, .{ .start = start, .end = i, .kind = .word });
            continue;
        }

        const seq_len = tui_text.utf8SeqLen(text[i]);
        i += if (i + seq_len <= text.len) seq_len else 1;
        try tokens.append(allocator, .{ .start = start, .end = i, .kind = .punct });
    }

    return tokens.toOwnedSlice(allocator);
}

fn markLcs(
    allocator: std.mem.Allocator,
    old_text: []const u8,
    old_tokens: []const Token,
    new_text: []const u8,
    new_tokens: []const Token,
    old_matched: []bool,
    new_matched: []bool,
) !void {
    const cols = new_tokens.len + 1;
    var dp = try allocator.alloc(usize, (old_tokens.len + 1) * cols);
    defer allocator.free(dp);
    @memset(dp, 0);

    var i = old_tokens.len;
    while (i > 0) {
        i -= 1;
        var j = new_tokens.len;
        while (j > 0) {
            j -= 1;
            const here = i * cols + j;
            if (tokenEqual(old_text, old_tokens[i], new_text, new_tokens[j])) {
                dp[here] = dp[(i + 1) * cols + (j + 1)] + 1;
            } else {
                dp[here] = @max(dp[(i + 1) * cols + j], dp[i * cols + (j + 1)]);
            }
        }
    }

    i = 0;
    var j: usize = 0;
    while (i < old_tokens.len and j < new_tokens.len) {
        if (tokenEqual(old_text, old_tokens[i], new_text, new_tokens[j])) {
            old_matched[i] = true;
            new_matched[j] = true;
            i += 1;
            j += 1;
        } else if (dp[(i + 1) * cols + j] >= dp[i * cols + (j + 1)]) {
            i += 1;
        } else {
            j += 1;
        }
    }
}

fn rangesFromTokens(allocator: std.mem.Allocator, tokens: []const Token, matched: []const bool) ![]Range {
    var ranges: std.ArrayList(Range) = .empty;
    errdefer ranges.deinit(allocator);

    var i: usize = 0;
    while (i < tokens.len) {
        if (matched[i]) {
            i += 1;
            continue;
        }

        const run_start = i;
        var has_significant = false;
        while (i < tokens.len and !matched[i]) : (i += 1) {
            if (tokens[i].kind != .whitespace) has_significant = true;
        }
        var start = run_start;
        var end = i;
        if (has_significant) {
            while (start < end and tokens[start].kind == .whitespace) : (start += 1) {}
            while (end > start and tokens[end - 1].kind == .whitespace) : (end -= 1) {}
        }
        if (start < end) try appendRange(allocator, &ranges, .{
            .start = tokens[start].start,
            .end = tokens[end - 1].end,
        });
    }

    return ranges.toOwnedSlice(allocator);
}

fn affixRanges(allocator: std.mem.Allocator, old_text: []const u8, new_text: []const u8) !PairRanges {
    const prefix = commonPrefix(old_text, new_text);
    const old_suffix = commonSuffix(old_text[prefix..], new_text[prefix..]);
    const new_suffix = old_suffix;

    var old_ranges: std.ArrayList(Range) = .empty;
    errdefer old_ranges.deinit(allocator);
    var new_ranges: std.ArrayList(Range) = .empty;
    errdefer new_ranges.deinit(allocator);

    if (prefix < old_text.len - old_suffix) try old_ranges.append(allocator, .{ .start = prefix, .end = old_text.len - old_suffix });
    if (prefix < new_text.len - new_suffix) try new_ranges.append(allocator, .{ .start = prefix, .end = new_text.len - new_suffix });

    const old_owned = try old_ranges.toOwnedSlice(allocator);
    errdefer allocator.free(old_owned);
    const new_owned = try new_ranges.toOwnedSlice(allocator);

    return .{
        .old = old_owned,
        .new = new_owned,
    };
}

fn appendRange(allocator: std.mem.Allocator, ranges: *std.ArrayList(Range), range: Range) !void {
    if (range.end <= range.start) return;
    if (ranges.items.len > 0 and ranges.items[ranges.items.len - 1].end >= range.start) {
        ranges.items[ranges.items.len - 1].end = @max(ranges.items[ranges.items.len - 1].end, range.end);
        return;
    }
    try ranges.append(allocator, range);
}

fn tokenEqual(left_text: []const u8, left: Token, right_text: []const u8, right: Token) bool {
    return std.mem.eql(u8, left_text[left.start..left.end], right_text[right.start..right.end]);
}

fn commonPrefix(left: []const u8, right: []const u8) usize {
    var i: usize = 0;
    var last_boundary: usize = 0;
    while (i < left.len and i < right.len and left[i] == right[i]) {
        i += 1;
        if (isUtf8Boundary(left, i) and isUtf8Boundary(right, i)) last_boundary = i;
    }
    return last_boundary;
}

fn commonSuffix(left: []const u8, right: []const u8) usize {
    var count: usize = 0;
    while (count < left.len and count < right.len and left[left.len - count - 1] == right[right.len - count - 1]) {
        count += 1;
    }
    while (count > 0 and (!isUtf8Boundary(left, left.len - count) or !isUtf8Boundary(right, right.len - count))) {
        count -= 1;
    }
    return count;
}

fn isUtf8Boundary(text: []const u8, index: usize) bool {
    return index == 0 or index == text.len or (text[index] & 0xc0) != 0x80;
}

fn isWhitespace(byte: u8) bool {
    return byte == ' ' or byte == '\t' or byte == '\r' or byte == '\n';
}

fn isWordByte(byte: u8) bool {
    return std.ascii.isAlphanumeric(byte) or byte == '_';
}

test "token diff keeps moved common run unhighlighted" {
    const allocator = std.testing.allocator;
    var ranges = try diffRanges(
        allocator,
        "    snapshot.files[state.active_file], store, view, state,",
        "    state, snapshot.files[state.active_file],",
    );
    defer ranges.deinit(allocator);

    try std.testing.expectEqual(@as(usize, 1), ranges.old.len);
    try std.testing.expectEqualStrings("store, view, state,", "    snapshot.files[state.active_file], store, view, state,"[ranges.old[0].start..ranges.old[0].end]);
    try std.testing.expectEqual(@as(usize, 1), ranges.new.len);
    try std.testing.expectEqualStrings("state,", "    state, snapshot.files[state.active_file],"[ranges.new[0].start..ranges.new[0].end]);
}

test "token diff marks only changed suffix" {
    const allocator = std.testing.allocator;
    var ranges = try diffRanges(allocator, "const name = oldValue;", "const name = newValue;");
    defer ranges.deinit(allocator);

    try std.testing.expectEqual(@as(usize, 1), ranges.old.len);
    try std.testing.expectEqualStrings("oldValue", "const name = oldValue;"[ranges.old[0].start..ranges.old[0].end]);
    try std.testing.expectEqual(@as(usize, 1), ranges.new.len);
    try std.testing.expectEqualStrings("newValue", "const name = newValue;"[ranges.new[0].start..ranges.new[0].end]);
}

test "token diff preserves whitespace-only changes" {
    const allocator = std.testing.allocator;
    var ranges = try diffRanges(allocator, "call(a, b)", "call(a,b)");
    defer ranges.deinit(allocator);

    try std.testing.expectEqual(@as(usize, 1), ranges.old.len);
    try std.testing.expectEqualStrings(" ", "call(a, b)"[ranges.old[0].start..ranges.old[0].end]);
    try std.testing.expectEqual(@as(usize, 0), ranges.new.len);
}

File: src/inline_diff.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Range {
    pub start: usize,
    pub end: usize,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct PairRanges {
    pub old: Vec<Range>,
    pub new: Vec<Range>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum TokenKind {
    Whitespace,
    Word,
    Punct,
}

#[derive(Clone, Copy, Debug)]
struct Token {
    start: usize,
    end: usize,
    kind: TokenKind,
}

const MAX_LCS_TOKENS: usize = 240;
const MAX_LCS_CELLS: usize = 24_000;

pub fn diff_ranges(old_text: &str, new_text: &str) -> PairRanges {
    if old_text == new_text {
        return PairRanges::default();
    }
    let old = tokenize(old_text);
    let new = tokenize(new_text);
    if old.is_empty()
        || new.is_empty()
        || old.len() > MAX_LCS_TOKENS
        || new.len() > MAX_LCS_TOKENS
        || old.len() * new.len() > MAX_LCS_CELLS
    {
        return affix_ranges(old_text, new_text);
    }

    let equal = |i: usize, j: usize| {
        old_text[old[i].start..old[i].end] == new_text[new[j].start..new[j].end]
    };
    let cols = new.len() + 1;
    let mut dp = vec![0usize; (old.len() + 1) * cols];
    for i in (0..old.len()).rev() {
        for j in (0..new.len()).rev() {
            dp[i * cols + j] = if equal(i, j) {
                dp[(i + 1) * cols + j + 1] + 1
            } else {
                dp[(i + 1) * cols + j].max(dp[i * cols + j + 1])
            };
        }
    }
    let mut old_matched = vec![false; old.len()];
    let mut new_matched = vec![false; new.len()];
    let (mut i, mut j) = (0, 0);
    while i < old.len() && j < new.len() {
        if equal(i, j) {
            old_matched[i] = true;
            new_matched[j] = true;
            i += 1;
            j += 1;
        } else if dp[(i + 1) * cols + j] >= dp[i * cols + j + 1] {
            // Keep the original deletion-first tie break for repeated tokens.
            i += 1;
        } else {
            j += 1;
        }
    }
    PairRanges {
        old: ranges_from_tokens(&old, &old_matched),
        new: ranges_from_tokens(&new, &new_matched),
    }
}

fn tokenize(text: &str) -> Vec<Token> {
    let mut tokens: Vec<Token> = Vec::new();
    for (start, ch) in text.char_indices() {
        let kind = if matches!(ch, ' ' | '\t' | '\r' | '\n') {
            TokenKind::Whitespace
        } else if ch.is_ascii_alphanumeric() || ch == '_' {
            TokenKind::Word
        } else {
            TokenKind::Punct
        };
        let end = start + ch.len_utf8();
        if let Some(previous) = tokens.last_mut()
            && kind != TokenKind::Punct
            && previous.kind == kind
        {
            previous.end = end;
        } else {
            tokens.push(Token { start, end, kind });
        }
        // Once over the cap, only the fact that the cap was exceeded matters.
        if tokens.len() > MAX_LCS_TOKENS {
            break;
        }
    }
    tokens
}

fn ranges_from_tokens(tokens: &[Token], matched: &[bool]) -> Vec<Range> {
    let mut ranges = Vec::new();
    let mut i = 0;
    while i < tokens.len() {
        if matched[i] {
            i += 1;
            continue;
        }
        let mut start = i;
        let mut significant = false;
        while i < tokens.len() && !matched[i] {
            significant |= tokens[i].kind != TokenKind::Whitespace;
            i += 1;
        }
        let mut end = i;
        if significant {
            while start < end && tokens[start].kind == TokenKind::Whitespace {
                start += 1;
            }
            while end > start && tokens[end - 1].kind == TokenKind::Whitespace {
                end -= 1;
            }
        }
        if start < end {
            append_range(
                &mut ranges,
                Range {
                    start: tokens[start].start,
                    end: tokens[end - 1].end,
                },
            );
        }
    }
    ranges
}

fn append_range(ranges: &mut Vec<Range>, range: Range) {
    if range.start >= range.end {
        return;
    }
    if let Some(previous) = ranges.last_mut()
        && previous.end >= range.start
    {
        previous.end = previous.end.max(range.end);
    } else {
        ranges.push(range);
    }
}

fn affix_ranges(old: &str, new: &str) -> PairRanges {
    let mut prefix = old
        .bytes()
        .zip(new.bytes())
        .take_while(|(a, b)| a == b)
        .count();
    while !old.is_char_boundary(prefix) || !new.is_char_boundary(prefix) {
        prefix -= 1;
    }
    let mut suffix = old[prefix..]
        .bytes()
        .rev()
        .zip(new[prefix..].bytes().rev())
        .take_while(|(a, b)| a == b)
        .count();
    while !old.is_char_boundary(old.len() - suffix) || !new.is_char_boundary(new.len() - suffix) {
        suffix -= 1;
    }
    let range = |text: &str| {
        if prefix < text.len() - suffix {
            vec![Range {
                start: prefix,
                end: text.len() - suffix,
            }]
        } else {
            Vec::new()
        }
    };
    PairRanges {
        old: range(old),
        new: range(new),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parts<'a>(text: &'a str, ranges: &[Range]) -> Vec<&'a str> {
        ranges
            .iter()
            .map(|range| &text[range.start..range.end])
            .collect()
    }

    #[test]
    fn token_diff_keeps_moved_common_run_unhighlighted() {
        let old = "    snapshot.files[state.active_file], store, view, state,";
        let new = "    state, snapshot.files[state.active_file],";
        let ranges = diff_ranges(old, new);
        assert_eq!(parts(old, &ranges.old), ["store, view, state,"]);
        assert_eq!(parts(new, &ranges.new), ["state,"]);
    }

    #[test]
    fn token_diff_marks_only_changed_suffix() {
        let old = "const name = oldValue;";
        let new = "const name = newValue;";
        let ranges = diff_ranges(old, new);
        assert_eq!(parts(old, &ranges.old), ["oldValue"]);
        assert_eq!(parts(new, &ranges.new), ["newValue"]);
    }

    #[test]
    fn token_diff_preserves_whitespace_only_changes() {
        let ranges = diff_ranges("call(a, b)", "call(a,b)");
        assert_eq!(parts("call(a, b)", &ranges.old), [" "]);
        assert!(ranges.new.is_empty());
        let ranges = diff_ranges(" \t", "  ");
        assert_eq!(parts(" \t", &ranges.old), [" \t"]);
        assert_eq!(parts("  ", &ranges.new), ["  "]);
    }

    #[test]
    fn multiple_edits_preserve_common_middle() {
        let old = "old(first, shared, last);";
        let new = "new(first, shared, end);";
        let ranges = diff_ranges(old, new);
        assert_eq!(parts(old, &ranges.old), ["old", "last"]);
        assert_eq!(parts(new, &ranges.new), ["new", "end"]);
    }

    #[test]
    fn identical_empty_insert_and_delete() {
        for text in ["", "a", "架😀e\u{301}"] {
            assert_eq!(diff_ranges(text, text), PairRanges::default());
        }
        assert_eq!(
            diff_ranges("", "架"),
            PairRanges {
                old: vec![],
                new: vec![Range { start: 0, end: 3 }]
            }
        );
        assert_eq!(
            diff_ranges("架", ""),
            PairRanges {
                old: vec![Range { start: 0, end: 3 }],
                new: vec![]
            }
        );
    }

    #[test]
    fn unicode_tokens_keep_byte_boundaries_and_common_characters() {
        let old = "call(架构, 😀);";
        let new = "call(架子, 😃);";
        let ranges = diff_ranges(old, new);
        assert_eq!(parts(old, &ranges.old), ["构", "😀"]);
        assert_eq!(parts(new, &ranges.new), ["子", "😃"]);
    }

    #[test]
    fn lcs_limits_use_utf8_safe_affix_ranges() {
        // 241 punctuation tokens exceed the token cap; 160 exceed the cell cap.
        for count in [160, MAX_LCS_TOKENS + 1] {
            let common = ";".repeat(count);
            let old = format!("{common}😀 middle À{common}");
            let new = format!("{common}😃 middle Ā{common}");
            let ranges = diff_ranges(&old, &new);
            assert_eq!(parts(&old, &ranges.old), ["😀 middle À"]);
            assert_eq!(parts(&new, &ranges.new), ["😃 middle Ā"]);
        }
    }

    #[test]
    fn affix_does_not_overlap_or_split_unicode() {
        for (old, new, old_parts, new_parts) in [
            ("😀", "😃", vec!["😀"], vec!["😃"]),
            ("À", "Ā", vec!["À"], vec!["Ā"]),
            ("a", "aa", vec![], vec!["a"]),
            ("架a", "架", vec!["a"], vec![]),
        ] {
            let ranges = affix_ranges(old, new);
            assert_eq!(parts(old, &ranges.old), old_parts);
            assert_eq!(parts(new, &ranges.new), new_parts);
        }
    }

    #[test]
    fn coalesce_only_overlapping_or_adjacent_nonempty_ranges() {
        let mut ranges = vec![];
        for (start, end) in [(1, 3), (2, 4), (4, 6), (9, 10), (11, 11)] {
            append_range(&mut ranges, Range { start, end });
        }
        assert_eq!(
            ranges,
            [Range { start: 1, end: 6 }, Range { start: 9, end: 10 }]
        );
    }

    #[test]
    fn generated_unicode_ranges_are_ordered_nonempty_and_in_bounds() {
        let atoms = [
            "", "a", "_", " ", "\t", "架", "😀", "\u{301}", ";", "old new",
        ];
        for a in atoms {
            for b in atoms {
                let old = format!("{a}({b}){a}");
                let new = format!("{b}({a}){b}");
                let pair = diff_ranges(&old, &new);
                for (text, ranges) in [(&old, &pair.old), (&new, &pair.new)] {
                    let mut previous = 0;
                    for range in ranges {
                        assert!(
                            previous <= range.start
                                && range.start < range.end
                                && range.end <= text.len()
                        );
                        assert!(
                            text.is_char_boundary(range.start) && text.is_char_boundary(range.end)
                        );
                        previous = range.end;
                    }
                }
            }
        }
    }
}
